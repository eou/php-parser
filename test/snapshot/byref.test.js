const parser = require('../main');

describe('byref', () => {
  it('variable', () => {
    expect(parser.parseEval('&$foo;')).toMatchSnapshot();
  });
  it('static lookup', () => {
    expect(parser.parseEval('&foo::$bar;')).toMatchSnapshot();
  });
  it('offset lookup', () => {
    expect(parser.parseEval('$foo[&$bar];')).toMatchSnapshot();
  });
  it('new class', () => {

    expect(
        parser.parseEval('$a =& new foo();', {
            parser: {
                php7: false
            }
        }
    )).toMatchSnapshot();
    expect(() => {
        parser.parseEval('$a =& new foo();');
    }).toThrow('Parse Error : syntax error, unexpected \'new\' (T_NEW) on line 1');
  });
  it('call result', () => {
    expect(parser.parseEval('$a =& foo( &$b );')).toMatchSnapshot();
  });
  it('return statement', () => {
    expect(parser.parseEval('return &$foo;')).toMatchSnapshot();
  });
  it('function definition', () => {
    expect(parser.parseEval('function &foo( &$bar ) { }')).toMatchSnapshot();
  });  
});
