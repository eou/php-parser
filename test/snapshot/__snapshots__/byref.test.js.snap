// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`byref call result 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": Variable {
          "curly": false,
          "kind": "variable",
          "name": "a",
        },
        "operator": "=",
        "right": ByRef {
          "kind": "byref",
          "what": Call {
            "arguments": Array [
              ByRef {
                "kind": "byref",
                "what": Variable {
                  "curly": false,
                  "kind": "variable",
                  "name": "b",
                },
              },
            ],
            "kind": "call",
            "what": ClassReference {
              "kind": "classreference",
              "name": "foo",
              "resolution": "uqn",
            },
          },
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref function definition 1`] = `
Program {
  "children": Array [
    _Function {
      "arguments": Array [
        Parameter {
          "byref": true,
          "kind": "parameter",
          "name": Identifier {
            "kind": "identifier",
            "name": "bar",
          },
          "nullable": false,
          "type": null,
          "value": null,
          "variadic": false,
        },
      ],
      "body": Block {
        "children": Array [],
        "kind": "block",
      },
      "byref": true,
      "kind": "function",
      "name": Identifier {
        "kind": "identifier",
        "name": "foo",
      },
      "nullable": false,
      "type": null,
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref new class 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": Variable {
          "curly": false,
          "kind": "variable",
          "name": "a",
        },
        "operator": "=",
        "right": ByRef {
          "kind": "byref",
          "what": New {
            "arguments": Array [],
            "kind": "new",
            "what": ClassReference {
              "kind": "classreference",
              "name": "foo",
              "resolution": "uqn",
            },
          },
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref offset lookup 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": OffsetLookup {
        "kind": "offsetlookup",
        "offset": ByRef {
          "kind": "byref",
          "what": Variable {
            "curly": false,
            "kind": "variable",
            "name": "bar",
          },
        },
        "what": Variable {
          "curly": false,
          "kind": "variable",
          "name": "foo",
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref return statement 1`] = `
Program {
  "children": Array [
    Return {
      "expr": ByRef {
        "kind": "byref",
        "what": Variable {
          "curly": false,
          "kind": "variable",
          "name": "foo",
        },
      },
      "kind": "return",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref static lookup 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": ByRef {
        "kind": "byref",
        "what": StaticLookup {
          "kind": "staticlookup",
          "offset": Variable {
            "curly": false,
            "kind": "variable",
            "name": "bar",
          },
          "what": ClassReference {
            "kind": "classreference",
            "name": "foo",
            "resolution": "uqn",
          },
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`byref variable 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": ByRef {
        "kind": "byref",
        "what": Variable {
          "curly": false,
          "kind": "variable",
          "name": "foo",
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;
